File: py/number-of-islands.py



''' DFS - visit each cell and mark adjacent land '''
class SolutionDFS:
    def numIslands(self, grid) -> int:
        if not grid: return 0

        island_count = 0
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == '1':
                    self.dfs(grid, i, j)
                    island_count += 1
        return island_count

    def dfs(self, grid, i, j):
        if i < 0 or j < 0 or i >= len(grid) or j >= len(grid[0]) or grid[i][j] != '1':
            return
        grid[i][j] = '#' # mark as visited

        # call dfs on adjacent squares
        self.dfs(grid, i-1, j)
        self.dfs(grid, i+1, j)
        self.dfs(grid, i, j-1)
        self.dfs(grid, i, j+1)



'''
Solution Explanation:
    Traverse the grid until we find a land block, then call DFS on any
    adjacent land blocks and mark them as "seen." Continue until we've called
    DFS for every land block.

If m and n are the number of rows and columns in the grid, respectively, then:
Time Complexity: O(m*n), since we may have to check every value in n for every
    value in m.
Space Complexity: O(m*n), in the worst case of all land blocks, we have to
    store the entire grid.
'''

class Solution0:
    def numIslands(self, grid) -> int:
        if not grid or not grid[0]:
            return 0

        rows = len(grid)
        columns = len(grid[0])
        count = 0

        for i in range(rows):
            for j in range(columns):
                # If we find a "1" (land block), run dfs at that location
                # in the grid.
                if '1' == grid[i][j]:
                    count += 1
                    self.dfs_iterative(grid, i, j)
        return count

    def inRange(self, grid: [[str]], r, c: [str]) -> bool:
        numRow, numCol = len(grid), len(grid[0])
        if r < 0 or c < 0 or r >= numRow or c >= numCol:
            return False
        return True

    def dfs_recursive(self, grid: [[str]], row, col: str) -> None:
        # Mark with a value of -1 if to declare that cell as "seen"
        grid[row][col] = -1

        # directions:  Up,     Down,    Left,    Right
        directions = [(0, 1), (0, -1), (-1, 0), (1, 0)]
        for d in directions:
            next_row = row + d[0]
            next_col = col + d[1]
            if self.inRange(grid, next_row, next_col) and \
               '1' == grid[next_row][next_col]:
                self.dfs_recursive(grid, next_row, next_col)


'''
Solution Explanation:
    Traverse the grid until we find a land block, then call BFS where we
    enqueues the current cell, marks it as seen, then searches all of its
    neighbors. If the queue is empty, then all lands blocks have been marked
    in the connected island.

If m and n are the number of rows and columns in the grid, respectively, then:
Time Complexity: O(m*n), since we may have to check every value in n for every
    value in m.
Space Complexity: O(min(m,n)), in the worst case where the entire grid is land
    blocks the queue will have at most min(m, n) elements.
'''


class Solution1:
    def inRange(self, grid: [[int]], r, c: [int]) -> bool:
        numRow, numCol = len(grid), len(grid[0])
        if r < 0 or c < 0 or r >= numRow or c >= numCol:
            return False
        return True

    def numIslands(self, grid: [[int]]) -> int:
        if not grid or not grid[0]:
            return 0

        rows = len(grid)
        columns = len(grid[0])
        count = 0

        for i in range(rows):
            for j in range(columns):
                # If we find a "1" (land block), run BFS at that location
                # in the grid.
                if "1" == grid[i][j]:
                    self.bfs(grid, i, j)
                    count += 1
        return count

    def bfs(self, grid: [[int]], row, col: int) -> None:
        queue = []
        queue.append((row, col))

        # Mark with a value of -1 if to declare that cell as "seen"
        grid[row][col] = -1

        while queue:
            # directions:  Up,     Down,    Left,    Right
            directions = [(0, 1), (0, -1), (-1, 0), (1, 0)]
            row, col = queue.pop()

            for d in directions:
                next_row = row + d[0]
                next_col = col + d[1]
                if self.inRange(grid, next_row, next_col) and "1" == grid[next_row][next_col]:
                    queue.append((next_row, next_col))
                    # Mark the cell as seen
                    grid[next_row][next_col] = -1

# Shorter pythonic solution
class Solution2:
    def numIslands(self, grid):
        def sink(i, j):
            # Bounds check
            if 0 <= i < len(grid) and 0 <= j < len(grid[i]) and grid[i][j] == '1':
                # Mark the cell as seen
                grid[i][j] = '#'

                # cast to list since map in python3 returns an iterator
                list(map(sink, (i+1, i-1, i, i), (j, j, j+1, j-1)))
                return 1
            return 0
        return sum(sink(i, j) for i in range(len(grid)) for j in range(len(grid[i])))


grid = [["1", "1", "1", "1", "0"],
        ["1", "1", "0", "1", "0"],
        ["1", "1", "0", "0", "0"],
        ["0", "0", "0", "0", "0"]]
print(Solution().numIslands(grid))
# 1

File: cpp/number-of-islands.cpp
// using DFS
class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
        if (grid.empty() or grid[0].empty()) {
            return 0;
        }

        int count = 0;
        for (int i = 0; i < grid.size(); ++i) {
            for (int j = 0; j < grid[0].size(); ++j) {
                if (grid[i][j] == '1') {
                    ++count;
                    dfs(grid, i, j);
                }
            }
        }
        return count;
    }

private:
    void dfs(vector<vector<char>>& grid, int i, int j) {
        // Mark the cell as seen
        grid[i][j] = '#';

        // Bounds check, if we find '1', call dfs
        // Check above/below
        if (i > 0 and grid[i-1][j] == '1') {
            dfs(grid, i-1, j);
        }
        if (i < grid.size()-1 and grid[i+1][j] == '1') {
            dfs(grid, i+1, j);
        }

        // Check left/right
        if (j > 0 and grid[i][j-1] == '1') {
            dfs(grid, i, j-1);
        }
        if (j < grid[0].size()-1 and grid[i][j+1] == '1') {
            dfs(grid, i, j+1);
        }
    }
};


// Using BFS
class Solution {
public:
    struct Point {
        int row, col;
        Point(int _x, int _y) : row(_x), col(_y) {}
    };

    int nr_row, nr_col;
    int numIslands(vector<vector<char> > &grid) {
        if (grid.empty() or grid[0].empty()) {
            return 0;
        }

        nr_row = grid.size();
        nr_col = grid[0].size();

        int count = 0;
        queue<Point> q;
        
        for (int i = 0; i < nr_row; ++i) {
            for (int j = 0; j < nr_col; ++j) {
                if (grid[i][j] == '0') {
                    continue;
                }
                ++count;

                // push the Point onto the queue
                grid[i][j] = '0';
                q.push(Point(i, j));
                while (!q.empty()) {
                    Point point = q.front();
                    q.pop();

                    // Enqueue all valid neighbords
                    add_neighbor(grid, point, -1,  0, q);
                    add_neighbor(grid, point,  0, -1, q);
                    add_neighbor(grid, point,  1,  0, q);
                    add_neighbor(grid, point,  0,  1, q);
                }
            }
        }
        return count;
    }

    void add_neighbor(vector<vector<char>>& grid, Point point, 
                      int d_row, int d_col, queue<Point> &q) {
        point.row += d_row;
        point.col += d_col;
        if (point.row >= nr_row || point.col >= nr_col || 
                point.row < 0 || point.col < 0 ||
                grid[point.row][point.col] == '0') {
            return;
        }
        grid[point.row][point.col] = '0';

        q.push(point);
    }
};


// using union-find (disjoint-set)
class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
        int sz = grid.size();
        int count = 0;
        vector<int> parent(sz ? sz * grid[0].size() : 0);
        for (int row = 0; row < grid.size(); ++row) {
            for (int col = 0; col < grid[row].size(); ++col) {
                if (grid[row][col] == '1') {
                    ++count;
                    int i = idx(row, col, sz);
                    parent[i] = i;
                    if (row > 0 && grid[row-1][col] == '1' && is_disjoint_set(i, idx(row-1, col, sz), parent))
                        --count;
                    if (col > 0 && grid[row][col-1] == '1' && is_disjoint_set(i, idx(row, col-1, sz), parent))
                        --count;
                }
            }
        }
        return count;
    }

    bool is_disjoint_set(int x, int y, vector<int>& parent) {
        int rootx = find(x, parent);
        int rooty = find(y, parent);
        if (rootx == rooty)
            return false;
        parent[rooty] = rootx;
        return true;
    }

    int find(int x, vector<int>& parent) {
        return (parent[x] == x) ? x : find(parent[x], parent);
    }
    
    inline int idx(int row, int col, int nrows) { return col * nrows + row; }
};
